Fix ubi OTA update

abgoyal opened the pull request:
Do not delete old volume, just overwrite its shadow and swap the names.

Delete won't work. Clarification - the delete ioctl itself will work, but then the new volume creation ioctl won't work due to the auto-resize volume taking up all available space.

abgoyal committed:
Fix ubi OTA update

Do not delete old volume, just overwrite its shadow and swap the names.

Delete won't work.

abgoyal committed:
Fix OTA update - 2

fix volume name comparison - match to provided name length, not to volume name length

abgoyal committed:
Fix confusing logging

reports writing to root, instead of root-backup. Young devs find that confusing :D

abgoyal committed:
Fix OTA update - 3 - the sigh edition

match lengths as well.

abgoyal committed:
Fix OTA update - the sigh edition

added more logging

wpt-nathan commented:
Couple of quick questions:
 - Why does delete not work? Can you elaborate on the issue?
 - What happens when UBI_IOCVOLUP is called with a size greater then the volume being overwritten?
 - I don't know if it still applies, but there was a dependency on the rootfs / _rootfs volume names in the init scripts.
 - buf ( line 494 ) may not always be '\0' terminated ( hence the original use of strncmp vs strcmp ), so strlen(buf) should just be len.

foxbunny commented:
I'll let @abgoyal answer the finer points. Our situation is:

- we have a single MTD partition that is used by 7 UBI volumes
- last UBI volume is auto-expanding
- when 2nd UBI volume (rootfs) is deleted by the installer, apparently the last one expands to take over the space, and there is no more space left for the new volume
- we have two volumes that are named `root` and `root-backup`, and the original string comparison code matched `root` when trying to match `root-backup`

abgoyal committed:
Fix volume name matching

strings read from /sys files may not always be zero-terminated. so can't use strlen.

abgoyal commented:
- adding to what Branko said, due to the auto-resize, adding a new volume doesn't work, which means old volume can not be deleted. 
- UBI_IOCVOLUP is being called on rootfs, which in our config is a fixed size volume. It cannot be larger that the size defined in our config, cause once again - auto-resize takes up all the rest of the MTD part. if at call time, size is greater than volume max size, call will fail. 
- did not know that the string from /sys won't always be zstr. Fixed. Curious as to how that happens?

wpt-nathan commented:
- I errored, it should actually be len -1 == strlen( name ). The sysfs entry returns 'chars\n' ( hence the strncmp len -1 ). It won't always be nul terminated because buf isn't memset( 0 ).

abgoyal committed:
Fix device name length off-by-one

/sys name entry contains "name/n", don't include /n in comparision.
